contracts/sales/FixedPriceSale.sol

```diff
@@ -95,240 +105,261 @@ contract FixedPriceSale {
+            TransferHelper.safeTransfer(
+                address(saleInfo.tokenIn),
+                owner,
+                saleStatus.tokensCommitted
+            );
+            uint256 soldTokens = _getTokenAmount(saleStatus.tokensCommitted);
+            uint256 unsoldTokens = uint256(saleInfo.tokensForSale).sub(
+                soldTokens
+            );
+            if (unsoldTokens > 0) {
                 TransferHelper.safeTransfer(
-                    address(tokenOut),
-                    _orderOwner,
-                    _purchasedTokens
+                    address(saleInfo.tokenOut),
+                    owner,
+                    unsoldTokens
                 );
             }
-            // delete last entry, even if tokensPurchased[_orderOwner] == 0 this okey, because then token has been claimed by claimTokens()
-            orderOwners.pop();
-            if (_counter == numberToDistributionPerBlock) {
-                break;
-            }
-            _counter++;
-        } // for
-        emit distributeAllTokensLeft(orderOwners.length);
+        } else {
+            TransferHelper.safeTransfer(
+                address(saleInfo.tokenOut),
+                owner,
+                saleInfo.tokensForSale.sub(
+                    _getTokenAmount(saleStatus.tokensCommitted)
+                )
+            );
+        }
+
+        emit SaleClosed();
     }
 
-    /// @dev count how many orders
-    function ordersCount() public view returns (uint256) {
-        return orderOwners.length;
+    /// @dev withdraws purchased tokens if sale successfull, if not releases committed tokens
+    function withdrawTokens(address user) public {
+        require(commitment[user] > 0, "FixedPriceSale: nothing to withdraw");
+        if (isMinRaiseReached()) {
+            require(saleStatus.isClosed, "FixedPriceSale: not closed yet");
+            uint256 withdrawAmount = _getTokenAmount(commitment[user]);
+            commitment[user] = 0;
+            TransferHelper.safeTransfer(
+                address(saleInfo.tokenOut),
+                user,
+                withdrawAmount
+            );
+            emit NewTokenWithdraw(user, withdrawAmount);
+        } else {
+            require(
+                block.timestamp > saleInfo.endDate,
+                "FixedPriceSale: endDate not reached"
+            );
+            uint256 releaseAmount = commitment[user];
+            commitment[user] = 0;
+            TransferHelper.safeTransfer(
+                address(saleInfo.tokenIn),
+                user,
+                releaseAmount
+            );
+            emit NewTokenRelease(user, releaseAmount);
+        }
     }
 
-    /// @dev withdraw collected funds
-    function _withdrawFunds() internal {
-        require(isClosed, "FixedPriceSale: sale not closed");
+    function _getTokenAmount(uint256 _amount) internal view returns (uint256) {
+        return _amount.mul(uint256(saleInfo.tokenPrice)).div(1e18);
```

Comment:
I'm not sure _getTokenAmount is correct

If the amount is 2 WXDAI
And token price is 4 WXDAI
Then the tokenAmount expected is 0.5 but this would give 8.
So shouldn't this be dividing? (2/4 = 0.5 tokenAmount)

This current method would work if price was in tokenOut but this is not what the name tokenPrice suggests to me.
Let me know if I missed something, but to me it seems tokenPrice name is misleading or it should be dividing.